### ISSUE TITLE
Smear and Stutter in VAT

### ISSUE THREAD

**Issue by Mohammad9760** [sharpen3d/openvat#4]
Hi, I created a test fracture simulation and created a VAT from it and it looked fine, so I made another fracture simulation with more pieces (higher vertex count) and more frames (longer simulation), and now the created VAT has some artifacts.

https://github.com/user-attachments/assets/1587a880-8e13-4b36-802a-caf2e43af0a5

I also tried it in game engine and the artifacts are more noticeable there.

**Comment by sharpen3d** [comment]
I have seen something similar before, when wrapping was not being calculated correctly with certain vertex counts - I believed to be resolved but obviously still has some issue. 

If you append your .blend I might be able to debug this specifically.

Generally if you keep your vertex count close to a power of 2 (16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384) and not **close but slightly over** - it has a better chance packing everything appropriately (same with frame counts).

This being said, it looks like this is a short amount of frames and not an insane amount of polygons - it may stem from something simple like mesh needing cleanup, vertex order changing during animation - but I would need to look at the file.


As far as being distorted in-engine moreso than blender - be sure you are importing the image as raw, 16bit (if you are using Unity and my importer this should be done for you - but it is curious that there is a difference between the blender preview and engine)

**Comment by sharpen3d** [comment]
the 'jitter' you are seeing at the end is simply the vat can only animate frame-by-frame whereas your source inherently interpolates between frames, as far as that goes, I would simulate at 0.5 speed then on playback play this vat at 2x speed in engine

**Comment by Mohammad9760** [comment]
> Generally if you keep your vertex count close to a power of 2 

noted, I'll try with different vert count.

> I would simulate at 0.5 speed then on playback play this vat at 2x speed in engine

makes perfect sense.

here's another test I did, the same simulation (same vert count) but 50 frames instead of 250:

https://github.com/user-attachments/assets/0529c7a2-1d04-447f-b407-b2f531a66d02


and here's the .blend file:
[fracture.blend.zip](https://github.com/user-attachments/files/18090798/fracture.blend.zip)


let me know if I can provide any other info to help.

**Comment by sharpen3d** [comment]
can you kindly include the .abc 'fragile_test_animated_rb.abc' referenced in the source's modifier?

**most likely** - the start and end position over 250 frames is too large to properly calculate small movement - it gets severely quantized because in RGB 16 bit you get 65,536 colors - which means ~65,536 unique positions. Usually this is OK, however I see in the calculated vat's modifier that it's range on Z is (0 to -461). This will make the blue channel have to account for all Z-positions over blue values (theoretically giving you minimum movement of (461/65536 = 0.00703430175) per frame on Z)

![vat_modifier](https://github.com/user-attachments/assets/7b67ff2a-8871-4afc-ae80-1c33fc6cf559)

**The best approach here is likely to limit the z maximum in your simulation** - (I'd fade out in a shader over z or something, but not sure your exact use case)

But, I'd still like to see the abc to see if there's any other weirdness happening and if there's something that can be improved on my end to account for situations like this.

**Comment by Mohammad9760** [comment]
> the start and end position over 250 frames is too large to properly calculate small movement
yep. the fractured pieces just keep falling down (I should've known better)

[fragile_test_animated_rb.abc.zip](https://github.com/user-attachments/files/18099470/fragile_test_animated_rb.abc.zip)


> The best approach here is likely to limit the z maximum in your simulation - (I'd fade out in a shader over z or something, but not sure your exact use case)

yeah good suggestion, it's for a game, some of the platforms are fragile, when you step on them they fall apart, I will stop the pieces from falling too far instead I'll scale them down when they fall down a bit farther.

**Comment by sharpen3d** [comment]
I should say by the way, this is a pretty cool use for the tool!

Leaving this unresolved until I check out the abc just to see if there's any weirdness in the script that can be solved or other recommendations

**Comment by Mohammad9760** [comment]
thanks for the help, here's what it looks right now in my game (totally a WIP)

https://github.com/user-attachments/assets/4206ec7d-608b-4321-b5b2-6a0027f515cc


by the way (unrelated here), I've seen fluid simulations baked into VATs (some Youtube video showing VATs in Unity made with Houdini), how can one make a VAT if the vertex count in variable throughout the animation?

**Comment by sharpen3d** [comment]
This is pretty cool by the way I like seeing the results!

To your question I missed when you posted:
There's some stuff I'm still unsure of when it comes to more advanced simulations but from what I know, a point cloud could be recorded, then 'wrapped' or 'skinned' in engine (do not know the proper term) in order to do things like fluid simulation. This would be like basically a mesh being shrink-wrapped over the points (in my understanding)

For variable vertices in general, currently you have to trick openvat into thinking it is constant. Basically it has to be constant domain size in verts, but does not matter how you get there. Order does matter as well. So if you can in any way keep the same amount of verts in the same order but change visibility (record another vat on the alpha chanel of the vat which corresponds to the vertex visibility or any other property). This is challenging though, requires you to edit the encoding nodes, then do some compositing...

I aspire to expand this toolset to be more similar to Houdini, but that's pretty much where my knowledge ends right now, what I understand is Houdini saves some extra special metadata to help with these scenarios.
